src/Components/WebAssembly/Authentication.Msal/src/Interop/AuthenticationService.ts

```diff
@@ -0,0 +1,363 @@
+    // state ourselves and pass an identifier to retrieve it while in the callback flow.
+    async saveState<T>(state: T): Promise<string> {
+        const base64UrlIdentifier = await new Promise<string>((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onloadend = evt => resolve((evt?.target?.result as string)
+                // The result comes back as a base64 string inside a dataUrl.
+                // We remove the prefix and convert it to base64url by replacing '+' with '-', '/' with '_' and removing '='.
+                .split(',')[1].replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, ''));
+            reader.onerror = evt => reject(evt.target?.error?.message);
+
+            // We generate a base 64 url encoded string of random data.
+            const entropy = window.crypto.getRandomValues(new Uint8Array(32));
+            reader.readAsDataURL(new Blob([entropy]));
+        });
+
+        sessionStorage.setItem(`${AuthenticationService._infrastructureKey}.AuthorizeService.${base64UrlIdentifier}`, JSON.stringify(state));
+        return base64UrlIdentifier;
+    }
+
+    async retrieveState<T>(url: string, isLogout: boolean = false): Promise<T | undefined> {
+        const parsedUrl = new URL(url);
+        const fromHash = parsedUrl.hash && parsedUrl.hash.length > 0 && new URLSearchParams(parsedUrl.hash.substring(1));
+        let state = fromHash && fromHash.getAll('state');
+        if (state && state.length > 1) {
+            return undefined;
+        } else if (!state || state.length == 0) {
+            state = parsedUrl.searchParams && parsedUrl.searchParams.getAll('state');
+            if (!state || state.length !== 1) {
+                return undefined;
+            }
+        }
+
+        // We need to calculate the state key in two different ways. The reason for it is that
+        // msal.js doesn't support the state parameter on logout flows, which forces us to shim our own logout state.
+        // The format then is different, as msal follows the pattern state=<<guid>>|<<user_state>> and our format
+        // simple uses <<base64urlIdentifier>>.
+        const appState = !isLogout? this._msalApplication.getAccountState(state[0]): state[0];
+        const stateKey = `${AuthenticationService._infrastructureKey}.AuthorizeService.${appState}`;
+        const stateString = sessionStorage.getItem(stateKey);
+        if (stateString) {
+            sessionStorage.removeItem(stateKey);
+            const savedState = JSON.parse(stateString);
+            return savedState;
+        }
+
+        return undefined;
+    }
+
+    purgeState() {
+        for (let i = 0; i < sessionStorage.length; i++) {
+            const key = sessionStorage.key(i);
+            if (key?.startsWith(AuthenticationService._infrastructureKey)) {
+                sessionStorage.removeItem(key);
+            }
+        }
```

Comment:
While I'm sure this works fine in most circumstances, this is `O(N)` for `N`="number of items in `sessionStorage`". There might be arbitrarily many things put there by other code.

Instead of putting everything in as separate top-level items, what about having a single top-level object `sessionStorage['Microsoft.Authentication.WebAssembly.Msal']`, and then store all our other values as properties on that object if it exists? Then the `purgeState` logic reduces to simply `delete sessionStorage['Microsoft.Authentication.WebAssembly.Msal']`.

Reply:
I don't think you are going to get thousands of entries in session storage. It essentially goes away when you close the tab. Other auth libraries do it this way (oidc-client, and msal with localStorage) so I think it is fine.

It did give me a bit of pause, but all libraries do it, so I'm not concerned.